core/arch/arm/kernel/thread_spmc_a64.S

```diff
@@ -171,13 +185,29 @@ END_FUNC thread_rpc
  * The current thread as indicated by @thread_index has just been
  * suspended.  The job here is just to inform normal world the thread id to
  * resume when returning.
+ * If the active FF-A endpoint is OP-TEE (or a TA) then an this function send an
+ * OPTEE_FFA_YIELDING_CALL_RETURN_INTERRUPT message to the normal world via the
```

Comment:
Is that a use case that someone calls OP-TEE or its TAs directly via the FF-A driver? If yes, the issue is already present without this patch, because OP-TEE currently sends `OPTEE_FFA_YIELDING_CALL_RETURN_INTERRUPT` in an `FFA_MSG_SEND_DIRECT_RESP` in case of a non-secure interrupt.
Until now the SPs were not preemptible by non-secure interrupts. This PR adds this and if an SP was running when the non-secure interrupt happened it sends `FFA_INTERRUPT`.
We still need to keep the original behavior for all other cases, because otherwise the OP-TEE driver won't be able to wake up thread waiters here: https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/tree/drivers/tee/optee/ffa_abi.c?h=v6.1.27#n593
Here I've relied on the active FF-A endpoint ID stored in rpc_target_info to distinguish between the two cases.

Reply:
> Here I've relied on the active FF-A endpoint ID stored in thread->tsd.rpc_target_info to distinguish between the two cases.

I'm not so keen on that since there is actually an overlap in case non-secure interrupts are unmasked at S-EL1. As long as OP-TEE core doesn't unmask non-secure interrupts while in S-EL1 while handling S-EL0 SPs it should still work. However, it's a bit fragile.

A new `THREAD_FLAGS_FFA_ONLY` flag to indicate that the normal OP-TEE MSG ABI isn't in use is more explicit. This can be used to stop any RPC while handling S-EL0 SPs.

Reply:
It should be added to `threads[n].state` when called from `thread_sp_alloc_and_run()`. I guess that means adding a `flags` parameter to `__thread_alloc_and_run()`.

Reply:
Ok, thank you, now I get it.
Done